data_structures/binary_trees/words_alphabetical.c

```diff
@@ -0,0 +1,241 @@
+char *getPointerToWord(char *word);
+struct Node *allocateMemoryForNode();
+void closeFile(FILE *file);
+void freeTreeMemory(struct Node *node);
+void endProgramAbruptly(char *errorMessage);
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main()
+{
+    struct Node *root = NULL;  ///< pointer to root node
+    FILE *file = NULL;         ///< pointer to file
+
+    file = fopen("file.txt", "r");
+    if (file == NULL)
+        endProgramAbruptly("file: 'file.txt' not found");
+
+    root = readWordsInFileToTree(file, root);
+    closeFile(file);
+
+    file = fopen("wordcount.txt", "a");
+    fprintf(file, "%-5s \t %9s \t %s \n", "S/N", "FREQUENCY", "WORD");
+    writeContentOfTreeToFile(root, file);
+    closeFile(file);
+
+    freeTreeMemory(root);
+
+    return 0;
+}
+
+/**
+ * @brief Read words from file to binary tree
+ * @param file file to be read from
+ * @param root root node of tree
+ * @return struct Node *
+ */
+struct Node *readWordsInFileToTree(FILE *file, struct Node *root)
+{
+    char *inputString =
+        (char *)malloc(46 * sizeof(char));  /// pointer to input string
+    // longest english word = 45 chars
+    // +1 for '\0' = 46 chars
+    char inputChar;                /// temp storage of characters
+    bool isPrevCharAlpha = false;  /// bool to mark the end of a word
+    int pos = 0;  /// position in inputString to place the inputChar
+    while ((inputChar = fgetc(file)) != EOF)
+    {
+        if (pos > 0)
+            isPrevCharAlpha = isalpha(inputString[pos - 1]);
+
+        // check if character is letter
+        if (isalpha(inputChar))
+        {
+            inputString[pos++] = tolower(inputChar);
+            continue;
+        }
+
+        // check if char is ' or - and if it is preceded by a letter
+        // eg yours-not, persons' (valid)
+        if ((inputChar == '\'' || inputChar == '-') && isPrevCharAlpha)
+        {
+            inputString[pos++] = inputChar;
+            continue;
+        }
+
+        // making sure that there is something valid in inputString
+        if (pos == 0)
+            continue;
+
+        // if last character is not letter and is not ' then replace by \0
+        if (!isPrevCharAlpha && inputString[pos - 1] != '\'')
+            pos--;
+        inputString[pos] = '\0';
+        pos = 0;
+        isPrevCharAlpha = false;
+        root = addWordToTree(inputString, root);
+    }
+    // this is to catch the case for the EOF
+    // being immediately after the last letter or '
+    if (pos > 0)
+    {
+        if (!isPrevCharAlpha && inputString[pos - 1] != '\'')
+            pos--;
+        inputString[pos] = '\0';
+        root = addWordToTree(inputString, root);
+    }
+
+    free(inputString);
+    return root;
+}
+
+/**
+ * @brief Adding words in to the correct position in the tree
+ * @param word word to be inserted in to the tree
+ * @param currentNode node which is being compared
+ * @return struct Node *
```

Comment:
You shouldn't provide what variable did it return. Same in other parts of the code (**see the example below**).

```c
 * @returns a pointer to the newly allocated node
```